### PR TITLE
Retry flashing on specific error conditions

### PR DESCRIPTION
**Pull request by juhhov**
## Status
**READY**

## Description
Retry flashing when it could be beneficial.

## Todos
- [x] Tests

### CODE REVIEW

**Comment by jonikula** [comment]
Just some pylint fixes to do there it seems.

File: mbed_flasher/erase.py [hunk 1]
```diff
@@ -234,12 +235,12 @@ def _erase_board_simple(self, target, no_reset):
         while auto_thread.is_alive():
             auto_thread.join(0.5)
 
-        new_target = MountVerifier(self.logger).check_points_unchanged(target)
-        if isinstance(new_target, int):
-            return new_target
+        target = FlasherMbed.refresh_target(target["target_id"])
```
Comment (YifeiZuo01) [inline mbed_flasher/erase.py]: Is that removing MountVerifier becuase mbed-ls has done such checks?
Does it mean we have to depend more on mbed-ls?

Comment (juhhov) [reply mbed_flasher/erase.py]: Yes, flasher should not need to do those checks.

File: mbed_flasher/common.py [hunk 2]
```diff
@@ -103,165 +97,24 @@ def __call__(self, name):
...
-            mounts = check_output(["mount", "|", "grep", "vfat"], shell=True).splitlines()
-            for mount in mounts:
-                if six.PY3:
-                    mount = mount.decode('utf-8')
-                if mount.find(new_target['dev_point']) != -1:
-                    new_target['mount_point'] = \
-                        mount.split('on')[1].split('type')[0].strip()
-                    return
-            sleep(1)
+        logger.info("Starting retry round {}".format(str(index + 1)))
```
Comment (jupe) [inline mbed_flasher/common.py]: log text is a bit weird

File: mbed_flasher/flash.py [hunk 3]
```diff
@@ -193,20 +193,15 @@ def flash_multiple(self, build, platform_name,
...
                              target_id=device['target_id'],
                              platform_name=None,
                              device_mapping_table=device_mapping_table,
                              method=method,
                              no_reset=no_reset)
-            if ret == EXIT_CODE_SUCCESS:
-                self.logger.debug("dev#%i -> SUCCESS", i)
-            else:
-                self.logger.warning("dev#%i -> FAIL", i)
+            if ret != EXIT_CODE_SUCCESS:
```
Comment (jupe) [inline mbed_flasher/flash.py]: this is actually breaking change: earlier we tried to flash all boards even some fails. Now we stop immediately when some fails. I'm fine with this change, but this need to be documented at least.

File: mbed_flasher/flashers/FlasherMbed.py [hunk 4]
```diff
@@ -239,34 +223,79 @@ def try_pyocd_flash(self, source, target):
...
+            if 'serial_port' in target and not no_reset:
+                self.reset_board(target['serial_port'])
+                sleep(0.1)
+
+            copy_file_success = self.copy_file(source, destination)
+            if copy_file_success == EXIT_CODE_FILE_COULD_NOT_BE_READ:
+                return EXIT_CODE_FILE_COULD_NOT_BE_READ
+
+            self.logger.debug("copy finished")
+            sleep(4)
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherMbed.py]: why this hard-coded delay? not good..

File: mbed_flasher/flashers/FlasherMbed.py [hunk 5]
```diff
@@ -239,34 +223,79 @@ def try_pyocd_flash(self, source, target):
...
+
+            thread = Thread(target=self.runner,
+                            args=([target['mount_point'], tail],))
+            thread.start()
+            while thread.is_alive():
+                thread.join(2.5)
+
+            if not no_reset:
+                self.reset_board(target['serial_port'])
+                sleep(0.4)
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherMbed.py]: why this hardcoded delay?

File: mbed_flasher/common.py [hunk 6]
```diff
@@ -103,165 +97,24 @@ def __call__(self, name):
         return self.logger
 
 
-class MountVerifier(object):
+def retry(logger, func, func_args, retries=DEFAULT_RETRY_AMOUNT, conditions=None):
```
[the remark this answers is not in the document]
Comment (juhhov) [reply mbed_flasher/common.py]: http://docs.python-guide.org/en/latest/writing/gotchas/

File: mbed_flasher/common.py (hunk 2 above)
Comment (juhhov) [reply mbed_flasher/common.py]: Suggestions?

File: mbed_flasher/flash.py (hunk 3 above)
Comment (juhhov) [reply mbed_flasher/flash.py]: I'll change this back.

File: mbed_flasher/flashers/FlasherMbed.py (hunk 4 above)
Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: Not in the scope of this PR and also other PR open for this (#83).

File: mbed_flasher/flashers/FlasherMbed.py (hunk 5 above)
Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: Not in the scope of thsi PR.

File: mbed_flasher/common.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonikula) [reply mbed_flasher/common.py]: Yep, no mutables as default values.

**Comment by juhhov** [comment]
No significant difference in flash durations.

v0.7.1    | This PR
------------|-----------
12.602s | 12.601s
13.599s | 12.631s
13.602s | 13.612s
12.636s | 12.613s
12.645s | 12.646s

**Pull request merged**